docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
+5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
+
+### Vision Beyond This Enhancement
+This proposal is part of a mid-term vision which enables the user to select an alternative ShootKapi scaling mode
+for any shoot. Said mode prioritises reliability, and is known to work well even under extreme API request profiles.
+
+The following approach is envisioned (discussion is limited to scaling ShootKapi):
+- The default scaling mode is the one proposed by this GEP.
+- For any shoot, the user is enabled to select an alternative scaling mode. In this mode, the user specifies a fixed
```

Comment:
I don't think we should allow users to interfere with the autoscaling method on a per-shoot basis.
Operators should make sure to gain confidence in the new mechanism and roll it out gradually (on a per-seed basis if desired).

Reply:
>I don't think we should allow users
Makes sense! We should not complicate user API in the absence of clear need. I changed "user" to "operator". Change will be available with next push.

>Operators should make sure to gain confidence in the new mechanism and roll it out gradually
That's exactly the plan. However, the new scaling method has known limitations and cases where it would predictably fail. So we need a reliable fallback. If some user has a use case where 90% of their Kapi requests are cluster-wide LIST (or get by label), I think it's better to have the option of operator setting a manual override, than being forced to say "we don't support your use case".

Reply:
>it's better to have the option of operator (...) override

@timebertt, if you agree with the above, I'll mark the conversation resolved.

Reply:
I don't really agree. We should strive towards a solution that also works for the extreme cases. There shouldn't be the need to overwrite things on a per-shoot basis.

Reply:
If operators notice problems with the new mechanism they should disable it on the entire seed/environment and we should improve the code.

Reply:
It seems that there is no room for discussion, so let's resolve this conversation.

Reply:
I don't think this is something we can even try to resolve on the abstract level of a GEP. I also think that we shouldn't built in things that we don't have evidence that we need it. So for now, I'd say let's drop the option to configure this on a per-Shoot-basis – regardless of _who_ is doing it (operator or Shoot owner, or someone else) – and wait for the cases where we really need this option.

For now, I think the only option we should build is to have the feature gate allowing to roll this out gradually one Seed after another.

Reply:
I hate to think about it like that, but we really don't have that many options. During design, we considered 7+ different scaling approaches. The one proposed here wasn't the best, functionally, but it was the best which fits within our resource constraints. We'll fine tune parameters and the signal driving HPA in the future, but other than that, algorithmic changes are off limits, unless we fork VPA/HPA.

Reply:
@oliver-goetz @timebertt @voelzmo 
The section discusses future vision, and is not meant to be part of this GEP's implementation. But to me, that was for reasons of optimal project execution, not because I doubt its usefulness.

Maybe I'm missing something. Here is my thinking:

_When the unsuitability of the new scaling approach becomes an issue for some customer cluster, that incident will be a here-and-now live service issue. Not a theoretical design objection. With the proposed fallback option, the time to accommodate the customer demand will be measured in minutes. On the other hand, updating Gardener design will take months._

_The parameter fine-tuning (requests/second per replica) will be governed by economy. We'll be optimising for the frequent case, intentionally leaving extreme outliers out. Taking into account the broad diversity of shoot loads supported by Gardener, I expect that there will be occasional outliers._

Anyway, I'm in the clear minority here, so it's impractical to drag the discussion any further. I'll remove that part of the document.